telemetry_package/main.py

```diff
@@ -1,34 +1,224 @@
+
+    def __rosout_callback(self, log_msg) -> None:
+        """
+        Callback function for the /rosout subscription.
+
+        This function takes each message that is NOT of the severity level DEBUG and sends it tho the ground-station webapp via UNIX socket.
+
+        Args:
+            log_msg: the incoming log message from the /rosout topic
+
+        Returns:
+            None
+        """
+
+        # Filter log messages
+        # Don't log DEBUG level messages
+        if log_msg.level == rclpy.logging.LoggingSeverity.DEBUG:
+            return
+        # Check if message should be throttled
+        log_is_higher_severity = log_msg.level == rclpy.logging.LoggingSeverity.WARN or log_msg.level == rclpy.logging.LoggingSeverity.ERROR or log_msg.level == rclpy.logging.LoggingSeverity.FATAL
+        if not log_is_higher_severity and self.throttle_log_message(log_msg):
+            return
+
+        try:
+            self.client_sock.sendall(json.dumps(
+                {"type": "log",
+                    "content": {
+                        "level": self.LOG_LEVELS[log_msg.level],
+                        "name": log_msg.name,
+                        "msg": log_msg.msg
+                    }}).encode() + b"\x17")
+        except Exception as e:
+            print(f'Error occurred in rosout_callback: {e}')
+
+    def __heartbeat_callback(self, hb_msg) -> None:
+        """
+        Callback function for the TopicNames.Heartbeat subscription. 
+
+        This function takes each message coming from TopicNames.Heartbeat and sends it to the ground-station webapp via UNIX socket.
+
+        Args: 
+            hb_msg: the incoming log message from the /rosout topic
+
+        Returns: 
+            None
+        """
+
+        try:
+            self.client_sock.sendall(json.dumps(
+                {"type": "heartbeat",
+                 "content": {
+                     "sender_id": hb_msg.sender_id,
+                     "tick": hb_msg.tick,
+                     "active": hb_msg.active,
+                     "timestamp": hb_msg.time_stamp.sec + hb_msg.time_stamp.nanosec / 1e9
+                 }}).encode() + b"\x17")
+        except Exception as e:
+            print(f'Error occurred in heartbeat_callback: {e}')
+
+    def __del__(self):
+        """ 
+        Destructor function of the TelemetryNode class. 
+
+        This function ensures that the server is closed cleanly and removed from disk. 
+
+        Args: 
+            None
+
+        Returns:
+            None 
+        """
+
+        print('Destructor called.')
+        self.server_sock.close()
+        os.remove(self.chosen_socket_path)
+
+
+def main(args=None):
+    rclpy.init(args=args)
+
+    telemetry_node = TelemetryNode()
+
+    rclpy.spin(telemetry_node)
+
+    telemetry_node.destroy_node()
+
+    # Ensures that the destructor function of the TelemetryNode object is executed on program shutdown
+    rclpy.on_shutdown(telemetry_node.__del__())

```

Comment:
I'm not sure, but shouldn't this line be in front of `spin`? How does ROS otherwise know that when it's spinning?

Reply:
Hi, which line are you referring to? ROS only needs to execute both functions after the node is done with `spin` , so I thought that they would come after. Also, sorry that I am replying this late, I missed seeing the pull request when it was published.

Reply:
Probably line 226. That looks weird to me.